### PR TITLE
Formula store and load tests

### PR DESCRIPTION
**Pull request by kquick**
Adds a framework for creating a Test Architecture with associated ParameterizedFormulas that are serialized to S-expression storage forms and then re-loaded and parsed to check for round-trip formula equivalence.

This is a minimal implementation with lots more that could be added, but its a (good?) start.

### CODE REVIEW

File: semmc/tests/semstore/ParamFormulaTests.hs [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+      property $ do Some r <- liftIO newIONonceGenerator
+                    sym <- liftIO $ newSimpleBackend r
+                    let opcode = OpPack
+                    (p, _operands) <- forAllT (genParameterizedFormula sym opcode)
+                    debugPrint $ "parameterizedFormula: " <> show p
+                    debugPrint $ "# literalVars: " <> show (MapF.size $ SF.pfLiteralVars p)
+                    debugPrint $ "# defs: " <> show (MapF.size $ SF.pfDefs p)
+                    let printedFormula = FO.printParameterizedFormula (HR.typeRepr opcode) p
+                    debugPrint $ "printedFormula: " <> show printedFormula
+                    let fenv = undefined
```
Comment (travitch) [inline semmc/tests/semstore/ParamFormulaTests.hs]: Minor nit: can we turn these `undefined` calls into `error` calls with a more descriptive message (e.g., "Formula parsing environment not supported in the test environment") in case something changes to trigger this later?

File: semmc/tests/semstore/ParamFormulaTests.hs [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+          let fenv = undefined
+          lcfg <- liftIO $ Log.mkLogCfg "rndtrip"
+          reForm <- liftIO $
+                    Log.withLogCfg lcfg $
+                    FI.readFormula sym fenv (HR.typeRepr opcode) printedFormula
+          debugPrint $ "re-Formulized: " <> show reForm
+          f <- evalEither reForm
+          -- verify the recreated formula matches the original
+          compareParameterizedFormulasSymbolically sym operands 1 p f
+
```
Comment (travitch) [inline semmc/tests/semstore/ParamFormulaTests.hs]: It is a bit hard to eyeball, but are these test bodies something that could be factored out into two or three shared functions? It is a bit hard to follow what is setup vs what is the main thing being tested - especially the three calls to `compareParameterizedFormulasSymbolically`.

File: semmc/tests/semstore/TestArch.hs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+  -- TBD: some memory locations
+  -- MemLoc :: Mem -> Location (BaseBVType 32)
+
+instance Show (TestLocation tp) where
+  show TestBarLoc     = "Bar"
+  show (TestBoxLoc n) = "Box_" <> show n
+  show (TestNatLoc n) = "NAT_" <> show n  -- KWQ: want NAT@... see above
+  show (TestIntLoc i) = if i >= 0
+                        then "INT_" <> show i  -- KWQ: want INT@... see above
+                        else "NEGINT_" <> show (-i)
```
Comment (travitch) [inline semmc/tests/semstore/TestArch.hs]: I wasn't able to figure out what the "see above" comment was referring to - are those strings already used for something else?

File: semmc/tests/semstore/TestArchPropGen.hs [hunk 4]
```diff
@@ -0,0 +1,625 @@
...
+               -> GenT m (WI.SymExpr sym (BaseBVType 32))
+genBV32SymExpr sym params opvars litvars = do
+  -- get a set of the possible variables that can be used as sources
+  -- in the defs, which are declared already as freshBoundVars.
+  let optional = catMaybes
+                 [
+                   case F.length params of
+                     0 -> Nothing
+                     _ -> Just (paramExprBV32 sym opvars litvars
+                                =<< HG.element (Set.toList params))
```
Comment (travitch) [inline semmc/tests/semstore/TestArchPropGen.hs]: Just trying to understand this - it chooses a parameter and builds a BV32 expr from it (where the parameter may refer to one of the litvars or opvars)?

Comment (travitch) [reply semmc/tests/semstore/TestArchPropGen.hs]: I think all of this looks good - I'm trying to wrap my head around what the pool of operands for generated expressions looks like - does it only include parameters, or also literals? If just parameters, how large do parameter lists in the generated formulas get?

File: semmc/tests/semstore/ParamFormulaTests.hs (hunk 2 above)
Comment (kquick) [reply semmc/tests/semstore/ParamFormulaTests.hs]: They can and should, and they mostly just need to be parameterized on opcode, which could then be another generated parameter, but the typing wasn't simple so I left this for later, but I think later is probably today.

File: semmc/tests/semstore/TestArch.hs (hunk 3 above)
Comment (kquick) [reply semmc/tests/semstore/TestArch.hs]: Whoops, I removed the original comment, which was on the TestBoxLoc show.  Basically it was a note that I originally had something like "Box#0", but that isn't a valid identifier (and for these it was going to be "NAT@{natval}").  I was initially questioning whether we needed to consider quoting these or alternatively detecting and rejecting them, but they should just be internal identifiers and we can control the use, so I think just removing the comments is the best route here.

File: semmc/tests/semstore/TestArchPropGen.hs (hunk 4 above)
Comment (kquick) [reply semmc/tests/semstore/TestArchPropGen.hs]: That's correct; the case statement + catMaybes is because the Hedgehog choice/element generators require a non-empty list.  If we have other optional elements they could be added here as well.

The parameters for an expression are comprised of a subset of the operands and an arbitrary number of arbitrarily generated literals.  There are currently no FunctionParameter values generated.

The debugPrint definition in TestUtils can be modified to have the tests print (lots of!) output about the generated stuff.  Here's an example of a formula for the OpSurf operator, which has 4 operands, and there were 3 literal variables generated as well for 2 formula definitions:

```
((operands
  ((user__9Vv58rJWt . 'Box)
   (user__L_WC_omh_PtJGOPMl3UDM2fE . 'Bar)
   (user__XMx4t__CwffO_4Uy_ . 'Box)
   (user__4nGak_7JBcgtPO_6xmkE . 'Box)))
  (in
   (user__9Vv58rJWt user__L_WC_omh_PtJGOPMl3UDM2fE user__XMx4t__CwffO_4Uy_ user__4nGak_7JBcgtPO_6xmkE 'Box_0 'Box_1 'Box_3))
  ((defs
   (('Box_0 #xb6a12ea4)
    ('Box_3
     (bvadd
      (bvneg
       (bvashr
        (bvudiv
         (bvudiv
          (bvand
           (bvmul
            (bvadd
             (bvadd #xffd70707 'Box_0)
             (bvneg
              (bvsdiv 'Box_1 user__XMx4t__CwffO_4Uy_)))
            (bvadd
             #xfe32d9d9
             (bvshl user__4nGak_7JBcgtPO_6xmkE #xff6386e6)))
           'Box_1)
          (bvxor
           (bvand
            (bvadd
             (bvand #xfefe5f8e 'Box_1)
             'Box_3)
            'Box_3)
           (bvshl
            (bvsdiv
             (bvor #xff2b01d5 user__9Vv58rJWt)
             #xffd5d5e9)
            (bvshl
             #xff8a9651
             (bvsdiv 'Box_3 #x00a12e6a)))))
         (bvor
          (bvadd
           (bvshl
            (bvneg
             (bvurem user__9Vv58rJWt #xffdd9b10))
            (bvashr
             (bvmul 'Box_1 'Box_0)
             (bvlshr 'Box_0 user__L_WC_omh_PtJGOPMl3UDM2fE)))
           (bvsdiv
            (bvor user__9Vv58rJWt #x0075e620)
            #xfe2b8ff2))
          (bvurem
           (bvnot
            (bvshl
             #x00000000
             (bvsdiv user__L_WC_omh_PtJGOPMl3UDM2fE #x00369396)))
           (bvudiv
            #xfb346d30
            (bvlshr
             (bvashr 'Box_3 #xfee01831)
             #xff1e1b48)))))
        (bvashr
         (bvashr
          (bvsrem
           (bvsdiv
            (bvadd
             #xc2cca010
             (bvneg
              (bvsdiv #xff329abe user__XMx4t__CwffO_4Uy_)))
            #xffffffff)
           #xfebaf9fd)
          #x00cd5bf9)
         (bvadd
          (bvadd
           #xff226e9e
           (bvashr
            (bvmul
             (bvmul #xfeffcb7c user__4nGak_7JBcgtPO_6xmkE)
             (bvlshr user__L_WC_omh_PtJGOPMl3UDM2fE 'Box_0))
            (bvlshr
             #xffffffff
             (bvmul user__XMx4t__CwffO_4Uy_ user__9Vv58rJWt))))
          (bvsdiv
           (bvmul
            (bvmul
             (bvsdiv user__XMx4t__CwffO_4Uy_ #xfffc00e1)
             (bvneg user__9Vv58rJWt))
            (bvadd
             #xff14174a
             (bvor user__XMx4t__CwffO_4Uy_ 'Box_1)))
           'Box_1)))))
      'Box_1))))))
```
